hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java

```diff
@@ -150,80 +157,88 @@ protected JpaDeploymentManagement(final EntityManager entityManager, final Actio
+        return assignDistributionSets(deploymentRequests, actionMessage, onlineDsAssignmentStrategy);
     }
 
-    @Override
-    public List<DistributionSetAssignmentResult> assignDistributionSets(final Set<Long> dsIDs,
-            final Collection<TargetWithActionType> targets) {
+    private List<DistributionSetAssignmentResult> assignDistributionSets(
+            final List<DeploymentRequest> deploymentRequests, final String actionMessage,
+            final AbstractDsAssignmentStrategy strategy) {
+        final List<DeploymentRequest> validatedRequests = validateRequestForAssignments(deploymentRequests);
+        final Map<Long, List<TargetWithActionType>> assignmentsByDsIds = convertRequest(validatedRequests);
 
-        final List<DistributionSetAssignmentResult> results = dsIDs.stream()
-                .map(dsID -> assignDistributionSetToTargets(dsID, targets, null, onlineDsAssignmentStrategy))
+        final List<DistributionSetAssignmentResult> results = assignmentsByDsIds.entrySet().stream()
+                .map(entry -> assignDistributionSetToTargetsWithRetry(entry.getKey(), entry.getValue(), actionMessage,
+                        strategy))
                 .collect(Collectors.toList());
-        onlineDsAssignmentStrategy.sendDeploymentEvents(results);
+        strategy.sendDeploymentEvents(results);
         return results;
     }
 
-    @Override
-    @Transactional(isolation = Isolation.READ_COMMITTED)
-    @Retryable(include = {
-            ConcurrencyFailureException.class }, maxAttempts = Constants.TX_RT_MAX, backoff = @Backoff(delay = Constants.TX_RT_DELAY))
-    public DistributionSetAssignmentResult assignDistributionSet(final long dsID,
-            final Collection<TargetWithActionType> targets, final String actionMessage) {
+    private List<DeploymentRequest> validateRequestForAssignments(List<DeploymentRequest> deploymentRequests) {
+        if (!isMultiAssignmentsEnabled()) {
+            deploymentRequests = deploymentRequests.stream().distinct().collect(Collectors.toList());
+            checkIfCausesMultiassignment(deploymentRequests);
+        }
+        checkQuotaForAssignment(deploymentRequests);
+        return deploymentRequests;
+    }
 
-        final DistributionSetAssignmentResult result = assignDistributionSetToTargets(dsID, targets, actionMessage,
-                onlineDsAssignmentStrategy);
-        onlineDsAssignmentStrategy.sendDeploymentEvents(result);
-        return result;
+    private static Map<Long, List<TargetWithActionType>> convertRequest(
+            final Collection<DeploymentRequest> deploymentRequests) {
+        return deploymentRequests.stream().collect(Collectors.groupingBy(DeploymentRequest::getDistributionSetId,
+                Collectors.mapping(DeploymentRequest::getTargetWithActionType, Collectors.toList())));
+    }
+
+    private static void checkIfCausesMultiassignment(final Collection<DeploymentRequest> deploymentRequests) {
```

Comment:
Semantics of this method seems to be to check if the deployment would actually require Multi-Assignments, right? If yes, I'd rename the method to sth like "checkIfRequiresMultiAssignment"...